SICP_book/chapter_2_1.metta

```diff
@@ -0,0 +1,793 @@
+    ($p4 (* (upper-bound $x) (upper-bound $y)))
+  )
+    (make-interval (min ($p1 $p2 $p3 $p4))
+                   (max ($p1 $p2 $p3 $p4)))))
+
+(= (compareList $curbest $list $f)
+    (if (== $list ())
+        $curbest
+        (let $cur (car-atom $list)
+            (let $tail (cdr-atom $list)
+                (if ($f $cur $curbest)
+                    (compareList $cur $tail $f)
+                    (compareList $curbest $tail $f))))))
+
+(= (min $varlist)
+    (compareList (car-atom $varlist) $varlist <))
+
+(= (max $varlist)
+    (compareList (car-atom $varlist) $varlist >))
+
+!(assertEqual
+    (max (5 2 4 6 1))
+    6)
+
+!(assertEqual
+    (min (5 2 4 6 1))
+    1)
+
+(= (div-interval $x $y)
+  (mul-interval $x
+                (make-interval (/ 1.0 (upper-bound $y))
+                               (/ 1.0 (lower-bound $y)))))
+
+(= (make-interval $a $b) ($a . $b))
+
+(= (upper-bound $interval) (let ($a . $b) $interval $b))
+(= (lower-bound $interval) (let ($a . $b) $interval $a))
+
+(= (r1) (make-interval 0.15 0.2))
+(= (r2) (make-interval 0.9 1.1))
+
+!(assertEqual
+    (add-interval (r1) (r2))
+    (1.05 . 1.3))
+
+!(assertEqual
+    (mul-interval (r1) (r2))
+    (0.135 . 0.22000000000000003))
+
+!(assertEqual
+    (div-interval (r1) (r2))
+    (0.13636363636363635 . 0.22222222222222224))
+; -----------------------End of Exercise 2.7----------------------------
+
+; Exercise 2.8.
+;
+; Using reasoning analogous to Alyssa's, describe how the difference of two
+; intervals may be computed. Define a corresponding subtraction procedure, called sub-interval.
+
+(= (sub-interval $x $y)
+    (make-interval (- (lower-bound $x) (lower-bound $y))
+                   (- (upper-bound $x) (upper-bound $y))))
+
+!(assertEqual
+    (sub-interval (r2) (r1))
+    (0.75 . 0.9000000000000001))
+
+; -----------------------End of Exercise 2.8----------------------------
+
+; Exercise 2.9.
+;
+; The width of an interval is half of the difference between its upper and lower bounds.
+; The width is a measure of the uncertainty of the number specified by the interval.
+; For some arithmetic operations the width of the result of combining two intervals
+; is a function only of the widths of the argument intervals, whereas for others the width
+; of the combination is not a function of the widths of the argument intervals. Show that
+; the width of the sum (or difference) of two intervals is a function only of the widths
+; of the intervals being added (or subtracted). Give examples to show that this is not true
+; for multiplication or division.
+
+(= (interval-width $interval)
+    (/ (- (upper-bound $interval) (lower-bound $interval)) 2))
+
+!(assertEqual
+    (interval-width (r2))
+    0.10000000000000003)
+
+; Well, width of sum-intervals indeed equal to sum of widths, but due to rounding inside metta I can't directly compare
+; results using assertEqual. But we will consider 0.125 and 0.12500000000000006 equal.
```

Comment:
It is possible to define:
```
(= (cmp-float $a $b)
    (let $diff (- $a $b)
        (if (< (Abs $diff) 1e-10) EQ
            (if (< $diff 0) LT GT) )))
```
And use it to compare float numbers.

Reply:
Thanks, I'll use this function.